### DISCUSSION TITLE
Use "mini buffer" panel instead of modal, for Search (`Cmd-F`)

### DISCUSSION THREAD

**Discussion by josdelien** [zed-industries/zed#24627] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

When searching (via `Cmd-F`) in a document, currently in Zed a search result modal opens with a way to toggle (using unbound `<` and `>` icon buttons with the mouse.) The problem with this UI/UX for document search is there's no quick overview, only an indication for the amount of search results found and a way to step over them.

A much faster and more insightful way (see screenshot) I've configured in Emacs Lisp (see screenshot) which I'd like to see implemented in Zed as well:

- a minibuffer is opened (at the bottom or top of the editor panel)
- **filtering** all instances of the search query
- that can be stepped via up/down arrow keys, and selected per <enter>
- that functions in parallel with the main document panel itself following the cursor.

This way, searching for common keywords (for example `fn` or `struct` in Rust, or `def` in Python) provides a very convenient and quick alternative to visual code folding or a mini-map to provide overview and quick intra-document navigation via keyboard bindings.

### If applicable, add mockups / screenshots to help present your vision of the feature

For example, in this `main.rs` (Rust) file,

The minibuffer search solution I envision (screenshot: my own Emacs configuration written in Emacs Lisp) for document search/filter...
![image](https://github.com/zed-industries/community/assets/8289122/e49c1306-0455-4b35-b94c-a964e3415acf)

... versus what it is now (same document, in Zed)
![image](https://github.com/zed-industries/community/assets/8289122/718d97f9-f6cb-4ff3-98d7-060206cee72a)

**Comment by josdelien** [comment]
Nota bene: 
what I've described as my intended default search UX (for `Cmd-x`), is currently already _somewhat implemented_ in the "Outline View" (`Cmd-Shift o`) modal, however that Outline view is limited to the LSP and file syntax tree.

==> Ideally I'd want to "merge" the current Search and Outline View functionality as the default Search behaviour, but not restricted to the current Outline View language syntax.

eg:
- `Cmd-F` should yield the current Outline View modal also bound to the underlying document buffer
- no pre-fetched outline results shown
- await a fuzzy search query, e.g. like in my screenshot from emacs `fn` as query to display all function definitions
- and then like right now in Outline View, be able to toggle via up/down arrow keys and hit <enter> to jump in there.

![image](https://github.com/zed-industries/community/assets/8289122/96f11121-639f-4b9b-a5a5-712b19d93041)

Nota bene: 
in this specific example, using the search query `fn`, Outline View happens to work. But let's say I'd for example use multiple comments throughout my code document with the word `// TODO: ...` this currently won't show in the Outline View.

**Comment by carlaviktor** [comment]
I've been using this in vim too with fzf. Use it all the time.

Another idea that I've been tinkering with for my setup is to adjust the weight of items depending on the proximity to where the cursor is.
